### ISSUE TITLE
jdk16 sanity.openjdk failures

### ISSUE THREAD

**Issue by pshipton** [eclipse/openj9#11135]
I ran a sanity.openjdk on jdk16 and found the following failures.

https://ci.eclipse.org/openj9/job/Test_openjdknext_j9_sanity.openjdk_s390x_linux_Personal/4

java/foreign/TestMemoryHandleAsUnsigned.java.TestMemoryHandleAsUnsigned
java/foreign/TestMismatch.java.TestMismatch
java/foreign/TestNoForeignUnsafeOverride.java.TestNoForeignUnsafeOverride

java/lang/invoke/MethodHandles/TestDropReturn.java.TestDropReturn
java/lang/invoke/MethodHandles/publicLookup/Driver.java.Driver
java/lang/invoke/MethodHandlesPermuteArgumentsTest.java.MethodHandlesPermuteArgumentsTest
java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java.SuperMethodTest

java/lang/reflect/sealed_classes/SealedClassesReflectionTest.java.SealedClassesReflectionTest

java/foreign/TestAdaptVarHandles.java.TestAdaptVarHandles
java/foreign/TestAddressHandle.java.TestAddressHandle
java/foreign/TestMemoryAccess.java.TestMemoryAccess
java/foreign/TestMemoryAlignment.java.TestMemoryAlignment
java/foreign/TestMemoryCopy.java.TestMemoryCopy
java/foreign/TestRebase.java.TestRebase
java/foreign/TestSegments.java.TestSegments
java/foreign/TestSharedAccess.java.TestSharedAccess
java/foreign/TestSlices.java.TestSlices
java/foreign/TestSpliterator.java.TestSpliterator
java/foreign/TestTypeAccess.java.TestTypeAccess
java/foreign/TestVarHandleCombinators.java.TestVarHandleCombinators
java/foreign/TestNative.java.TestNative

java/util/stream/test/org/openjdk/tests/java/util/stream/SegmentTestDataProvider.java.SegmentTestDataProvider
java/util/stream/test/org/openjdk/tests/java/util/stream/SpliteratorTest.java.SpliteratorTest

**Comment by pshipton** [comment]
@tajila @babsingh fyi

**Comment by babsingh** [comment]
Acknowledged. Will add these failures to my TODO list.

**Comment by babsingh** [comment]
All **"java/foreign/*"** testcases are failing because of the below exception:
```
15:05:19  java.lang.ClassCastException: java.lang.invoke.DirectHandle incompatible with java.lang.invoke.BoundMethodHandle
15:05:19  	at java.base/java.lang.invoke.VarHandles.noCheckedExceptions(VarHandles.java:625)
...
```

**FIX (impacts the JDK-next extension's repo):** https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/248. With this **FIX**, `make test-jdk_foreign` passes locally for me. 

The above **FIX** will allow OpenJDK's `VarHandles` and `IndirectVarHandle` to work with OpenJ9's `VarHandle` implementation.

But, the above **FIX** will need to be reverted for testing the new OpenJDK JSR292 support (#7352). 

@tajila @pshipton Should we merge the above **FIX** to resolve the failures, and then revert it once OpenJDK JSR292 support is complete?

**Comment by babsingh** [comment]
**While building JDK16, Java warnings are being treated as errors:**
```
*/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavaTokenizer.java:1059: warning: [removal] stripIndent() in String has been deprecated and marked for removal
                        string = string.stripIndent();
                                       ^
*/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavaTokenizer.java:1068: warning: [removal] translateEscapes() in String has been deprecated and marked for removal
                        string = string.translateEscapes();
```

**Temporary FIX (impacts the JDK-next extension's repo):**
```
diff --git a/make/common/JavaCompilation.gmk b/make/common/JavaCompilation.gmk
index bedb971..47d08b2 100644

-  JAVA_WARNINGS_ARE_ERRORS ?= -Werror
+#  JAVA_WARNINGS_ARE_ERRORS ?= -Werror
```

**Comment by babsingh** [comment]
> **java/lang/invoke/MethodHandles/TestDropReturn.java.TestDropReturn**

A new method, `MethodHandles.dropReturn`, needs to be implemented for OpenJ9 for JDK16:
```
// JDK16+
// Replace the return type of "mh" with a void return type
// Throw NPE if mh is null
public static MethodHandle dropReturn(MethodHandle mh) {
    return mh with void return type;
}
```

If OpenJDK JSR292 support (#7352) is complete for JDK16, then the above work won't be needed.

**Comment by babsingh** [comment]
## Remaining Failures: Pending Investigation

> java/lang/invoke/MethodHandles/publicLookup/Driver.java.Driver

```
14:41:37  CL1: verify access for [java.lang.Object/publicLookup, q.EndPoint/publicLookup, q.EndPoint/publicLookup]
14:41:37  java.lang.AssertionError: unexpected result
14:41:37  	at m1/p.Main.assertTrue(Main.java:104)
14:41:37  	at m1/p.Main.checkTypeConsistency(Main.java:98)
14:41:37  	at m1/p.Main$Test.verifyAccess(Main.java:66)
14:41:37  	at m1/p.Main.main(Main.java:89)
...
```

> java/lang/invoke/MethodHandlesPermuteArgumentsTest.java.MethodHandlesPermuteArgumentsTest
```
14:45:19  JavaTest Message: JUnit Failure: testPermuteArguments(test.java.lang.invoke.MethodHandlesPermuteArgumentsTest): null
14:45:19  java.lang.NullPointerException
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest.assertMatches(MethodHandlesPermuteArgumentsTest.java:248)
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest.assertThrows(MethodHandlesPermuteArgumentsTest.java:240)
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest.testBadReorderArrayLength(MethodHandlesPermuteArgumentsTest.java:203)
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest.testPermuteArguments0(MethodHandlesPermuteArgumentsTest.java:62)
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest$$Lambda$9/0x00000000bb166510.run(Unknown Source)
14:45:19  	at test.java.lang.invoke.lib.CodeCacheOverflowProcessor$$Lambda$10/0x00000000bb166c10.run(Unknown Source)
14:45:19  	at jdk.test.lib.Utils.filterException(Utils.java:633)
14:45:19  	at test.java.lang.invoke.lib.CodeCacheOverflowProcessor.runMHTest(CodeCacheOverflowProcessor.java:71)
14:45:19  	at test.java.lang.invoke.MethodHandlesPermuteArgumentsTest.testPermuteArguments(MethodHandlesPermuteArgumentsTest.java:51)
```

> java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java.SuperMethodTest
```
14:41:37  test p.SuperMethodTest.splitPackage(): failure
14:41:37  java.lang.IllegalAccessError: Class p/SuperMethodTest$B$$Lambda$10/0x000000003c048ce0 illegally accessing "protected" member of class p/SuperMethodTest$A
14:41:37  	at p.SuperMethodTest$B$$Lambda$10/0x000000003c048ce0.run(Unknown Source)
14:41:37  	at p.SuperMethodTest.splitPackage(SuperMethodTest.java:101)
```

> java/lang/reflect/sealed_classes/SealedClassesReflectionTest.java.SealedClassesReflectionTest
```
//Seen 8 times
14:46:44  test SealedClassesReflectionTest.testSealedClasses(java.lang.Class@16c8b04a): failure
14:46:44  java.lang.AssertionError: expected [true] but found [false]
14:46:44  	at org.testng.Assert.fail(Assert.java:94)
14:46:44  	at org.testng.Assert.failNotEquals(Assert.java:496)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:42)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:52)
14:46:44  	at SealedClassesReflectionTest.testSealedClasses(SealedClassesReflectionTest.java:87)

//Seen 9 times
14:46:44  test SealedClassesReflectionTest.testSealedReflection(java.lang.Class@16c8b04a, 2, [Ljava.lang.String;@f4f71042, [Ljava.lang.Class;@503931f0, [LSealedClassesReflectionTest$SealedStatus;@5f672dcb): failure
14:46:44  java.lang.AssertionError: expected [true] but found [false]
14:46:44  	at org.testng.Assert.fail(Assert.java:94)
14:46:44  	at org.testng.Assert.failNotEquals(Assert.java:496)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:42)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:52)
14:46:44  	at SealedClassesReflectionTest.testSealedReflection(SealedClassesReflectionTest.java:217)

//Seen 3 times
14:46:44  test SealedClassesReflectionTest.testnon_sealedClasses(java.lang.Class@3b2700e3): failure
14:46:44  java.lang.AssertionError: expected [true] but found [false]
14:46:44  	at org.testng.Assert.fail(Assert.java:94)
14:46:44  	at org.testng.Assert.failNotEquals(Assert.java:496)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:42)
14:46:44  	at org.testng.Assert.assertTrue(Assert.java:52)
14:46:44  	at SealedClassesReflectionTest.testnon_sealedClasses(SealedClassesReflectionTest.java:130)
```

**Comment by pshipton** [comment]
> Should we merge the above FIX to resolve the failures, and then revert it once OpenJDK JSR292 support is complete

From my point of view, these tests don't need to be passing atm, the failures don't block any promotion. I'll defer to Tobi. If we think OpenJDK JSR292 support will be complete for JDK16, I'm happy to wait for that to be enabled by default.

**Comment by babsingh** [comment]
**SUMMARY:** All failures reported here except the **SealedClassesReflectionTest** failures should be resolved automatically, once OpenJDK JSR292 support (#7352) is complete for JDK16.

**Comment by tajila** [comment]
The required changes don't seem like it will be a big effort for us. Based on this, and on the fact that we do not need these tests to be passing right now I think we should revisit this issue in January 2021. By then OpenJDK  MH work should be functional meaning that support this will be no/little effort. If in January we feel that #7352 is not ready, we will still have time to provide a J9 MH implementation.

**Comment by JasonFengJ9** [comment]
Note that is `SealedClassesReflectionTest` is tracked by https://github.com/eclipse/openj9/issues/10581 as well.

**Comment by babsingh** [comment]
> Note that `SealedClassesReflectionTest` is tracked by #10581 as well.

#10581 only covers the `SealedClassesReflectionTest.java:87` failure. There are two other failure types: `SealedClassesReflectionTest.java:217` and `SealedClassesReflectionTest.java:130`. Refer to https://github.com/eclipse/openj9/issues/11135#issuecomment-724831028.

**Comment by JasonFengJ9** [comment]
Excluding following tests via https://github.com/AdoptOpenJDK/openjdk-tests/pull/2088

```
java/lang/invoke/MethodHandles/publicLookup/Driver.java
java/lang/invoke/MethodHandlesPermuteArgumentsTest.java
java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java
java/lang/invoke/MethodHandles/TestDropReturn.java
java/util/stream/test/org/openjdk/tests/java/util/stream/SpliteratorTest.java
```

They are expected to be fixed by OpenJDK JSR292 support (#7352) for JDK16 as per https://github.com/eclipse/openj9/issues/11135#issuecomment-724831028 & https://github.com/eclipse/openj9/issues/11135#issuecomment-724821524.

`java/util/stream/test/org/openjdk/tests/java/util/stream/SpliteratorTest.java` wasn't mentioned here specifically though it failed in a similar way, so I assume it can be fixed along with others.
@babsingh please comment if otherwise.

fyi @tajila @pshipton

**Comment by JasonFengJ9** [comment]
Summary of other issues for JDK16 test suite `sanity.openjdk`:

* JDK16 InternalError: Compile stub invoked! at java.base/java.lang.invoke.VarHandle.withInvokeBehavior(VarHandle.java:1675) 
https://github.com/eclipse/openj9/issues/11256
 
* JDK16 java/lang/invoke/lambda/superProtectedMethod/InheritedProtectedMethod.java IllegalAccessError: illegally accessing "package private" member of class java/lang/invoke/MethodHandles
https://github.com/eclipse/openj9/issues/11359

* JDK16 java/foreign/TestCleaner.java j9jcl.347 * ** ASSERTION FAILED ** at common/jdk_internal_misc_ScopedMemoryAccess.cpp:40: ((0 ))
https://github.com/eclipse/openj9/issues/11360

* JDK16 : java/lang/invoke/MethodHandles/classData/ClassDataTest.java error: <T>classData(Lookup,String,Class<T>) is not public in MethodHandles; cannot be accessed from outside package
https://github.com/eclipse/openj9/issues/11366

**Comment by babsingh** [comment]
> `java/util/stream/SpliteratorTest.java` wasn't mentioned here specifically

It is mentioned in https://github.com/eclipse/openj9/issues/11135#issuecomment-724814903: **java/util/stream/test/\*** (entire package).

**Comment by keithc-ca** [comment]
I find if you use a recent jdk15 (e.g. from adoptopenjdk.net) for your bootjdk, the warnings about preview features do not occur.

**Comment by pshipton** [comment]
Latest list of failures with the latest jdk16 content (ignoring crypto failures which have another issue).

https://ci.eclipse.org/openj9/job/Test_openjdk16_j9_sanity.openjdk_s390x_linux_Nightly/1/

java/lang/invoke/MethodHandles/classData/ClassDataTest.java.ClassDataTest
java/lang/invoke/lambda/LambdaReceiver.java.LambdaReceiver
java/lang/invoke/lambda/LambdaReceiverBridge.java.LambdaReceiverBridge
java/lang/invoke/lambda/superProtectedMethod/InheritedProtectedMethod.java.InheritedProtectedMethod
java/lang/reflect/records/IsRecordTest.java.IsRecordTest
java/lang/reflect/sealed_classes/SealedClassesReflectionTest.java.SealedClassesReflectionTest

**Comment by tajila** [comment]
https://github.com/eclipse/openj9/issues/11603

**Comment by tajila** [comment]
The SealedClassesReflectionTest failures are caused by #10977

**Comment by tajila** [comment]
Opened https://github.com/eclipse/openj9/issues/11605 to track:

java/lang/invoke/lambda/LambdaReceiver.java.LambdaReceiver
java/lang/invoke/lambda/LambdaReceiverBridge.java.LambdaReceiverBridge
java/lang/invoke/lambda/superProtectedMethod/InheritedProtectedMethod.java.InheritedProtectedMethod

**Comment by tajila** [comment]
The failures below are all related:

java/lang/invoke/MethodHandles/classData/ClassDataTest.java.ClassDataTest

java/lang/invoke/lambda/LambdaReceiver.java.LambdaReceiver
java/lang/invoke/lambda/LambdaReceiverBridge.java.LambdaReceiverBridge
java/lang/invoke/lambda/superProtectedMethod/InheritedProtectedMethod.java.InheritedProtectedMethod

being tracked in https://github.com/eclipse/openj9/issues/11366

**Comment by pshipton** [comment]
This is no longer tracking anything useful (if it ever was).

**Comment by pshipton** [comment]
Hmm, although there are some tests excluded under this Issue.
https://github.com/eclipse/openj9/issues/11135#issuecomment-738920269

@babsingh are there PRs that will fix the tests excluded in https://github.com/AdoptOpenJDK/openjdk-tests/pull/2088 under this issue?

**Comment by babsingh** [comment]
@pshipton Yes, the following issues are still unresolved:

1. java/lang/invoke/MethodHandles/classData/ClassDataTest.java (being tracked via https://github.com/eclipse/openj9/issues/11366)
2. java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java (no tracker found)
3. java/lang/invoke/MethodHandles/publicLookup/Driver.java (being tracked via https://github.com/eclipse/openj9/issues/10648)
4. java/lang/reflect/sealed_classes/SealedClassesReflectionTest.java (should be fixed by https://github.com/eclipse/openj9/pull/11606)
5. java/lang/invoke/VarHandles/VarHandleTestExact.java (being tracked via https://github.com/eclipse/openj9/issues/11256)

**Notes:**
- I am current working (1) which is the most important one. 
- Not sure if (2) and (3) will be needed for JDK16; but, they will need to assigned to someone. 
- (4) should be fixed once https://github.com/eclipse/openj9/pull/11606 is merged.
- (5) seems related to the Foreign Memory Access API; it may not be needed for JDK16.

**Comment by babsingh** [comment]
@pshipton All important issues, which are reported here, have been addressed for the 0.25 release. Any unresolved issue has its own standalone tracker. So, we can close this issue.

**Comment by pshipton** [comment]
We need to fix the excludes to refer to the correct issues/PRs before closing this. How can we re-assign the following?

java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java
java/lang/invoke/MethodHandlesPermuteArgumentsTest.java

**Comment by babsingh** [comment]
>We need to fix the excludes to refer to the correct issues/PRs before closing this. How can we re-assign the following?

- Opened https://github.com/AdoptOpenJDK/openjdk-tests/pull/2215 to remove references of this PR from the exclusion list.
- Opened https://github.com/eclipse/openj9/issues/11822 for `MethodHandlesPermuteArgumentsTest.java`.
